src/dodal/devices/i22/dcm.py

```diff
@@ -0,0 +1,152 @@
+import time
+from dataclasses import dataclass
+from typing import Dict, Literal
+
+from bluesky.protocols import Reading
+from event_model.documents.event_descriptor import DataKey
+from ophyd_async.core import ConfigSignal, StandardReadable, soft_signal_r_and_setter
+from ophyd_async.epics.motion import Motor
+from ophyd_async.epics.signal import epics_signal_r
+
+# Conversion constant for energy and wavelength, taken from the X-Ray data booklet
+# Converts between energy in KeV and wavelength in angstrom
+_CONVERSION_CONSTANT = 12.3984
+
+
+@dataclass(frozen=True, unsafe_hash=True)
+class CrystalMetadata:
+    """
+    Metadata used in the NeXus format,
+    see https://manual.nexusformat.org/classes/base_classes/NXcrystal.html
+    """
+
+    usage: Literal["Bragg", "Laue"] | None = None
+    type: str | None = None
+    reflection: tuple[int, int, int] | None = None
+    d_spacing: float | None = None
+
+
+class DoubleCrystalMonochromator(StandardReadable):
+    """
+    A double crystal monochromator (DCM), used to select the energy of the beam.
+
+    perp describes the gap between the 2 DCM crystals which has to change as you alter
+    the angle to select the requested energy.
+
+    offset ensures that the beam exits the DCM at the same point, regardless of energy.
+    """
+
+    def __init__(
+        self,
+        motion_prefix: str,
+        temperature_prefix: str,
+        crystal_1_metadata: CrystalMetadata | None = None,
+        crystal_2_metadata: CrystalMetadata | None = None,
+        prefix: str = "",
+        name: str = "",
+    ) -> None:
+        with self.add_children_as_readables():
+            # Positionable Parameters
+            self.bragg = Motor(motion_prefix + "BRAGG")
+            self.offset = Motor(motion_prefix + "OFFSET")
+            self.perp = Motor(motion_prefix + "PERP")
+            self.energy = Motor(motion_prefix + "ENERGY")
+            self.crystal_1_roll = Motor(motion_prefix + "XTAL1:ROLL")
+            self.crystal_2_roll = Motor(motion_prefix + "XTAL2:ROLL")
+            self.crystal_2_pitch = Motor(motion_prefix + "XTAL2:PITCH")
+
+            # Temperatures
+            self.backplate_temp = epics_signal_r(float, temperature_prefix + "PT100-7")
+            self.perp_temp = epics_signal_r(float, temperature_prefix + "TC-1")
+            self.crystal_1_temp = epics_signal_r(float, temperature_prefix + "PT100-1")
```

Comment:
nit: Ideally, temperature would be a readable on crystal_N, so the reading came in as "dcm-crystal_N-temperature" rather than "dcm-crystal_N_temperature".

Other crystal fields could be on the crystal, if this is i22 specific impl.